subsystems/armsubsystem.py

```diff
@@ -0,0 +1,830 @@
+        )
+
+        if not Preferences.getBoolean(constants.kArmObeyEndstopsKey, True):
+            clampedShoulder = optimizeAngle(
+                Rotation2d.fromDegrees(90), shoulder
+            ).radians()
+            clampedElbow = angleModulus(elbow.radians())
+            clampedWrist = angleModulus(wrist.radians())
+
+        trueShoulderPos = clampedShoulder
+        trueElbowPos = clampedElbow + trueShoulderPos
+        trueWristPos = clampedWrist + trueElbowPos
+
+        desiredInterpolation: ArmSubsystem.InterpolationMethod = (
+            self.interpolationMethod.getSelected()
+        )
+
+        currentWristRaw = self._getWristRawArmRotation()
+        if desiredInterpolation == ArmSubsystem.InterpolationMethod.JointSpace:
+            currentElbowRaw = self._getWristRawArmRotation()
+            currentShoulderRaw = self._getShoulderRawArmRotation()
+
+            trueShoulderPos = (
+                self.shoulderPID.calculate(
+                    currentShoulderRaw.radians(), trueShoulderPos
+                )
+                + currentShoulderRaw.radians()
+            )
+            trueElbowPos = (
+                self.elbowPID.calculate(currentElbowRaw.radians(), trueElbowPos)
+                + currentElbowRaw.radians()
+            )
+
+        trueWristPos = (
+            self.thetaProfiledPID.calculate(currentWristRaw.radians(), trueWristPos)
+            + currentWristRaw.radians()
+        )
+
+        SmartDashboard.putNumber(
+            constants.kArmShoulderTargetMotorKey,
+            trueShoulderPos / constants.kRadiansPerDegree,
+        )
+        SmartDashboard.putNumber(
+            constants.kArmElbowTargetMotorKey,
+            trueElbowPos / constants.kRadiansPerDegree,
+        )
+        SmartDashboard.putNumber(
+            constants.kArmWristTargetMotorKey,
+            trueWristPos / constants.kRadiansPerDegree,
+        )
+
+        shoulderArmEncoderPulses = (
+            trueShoulderPos
+            * constants.kTalonEncoderPulsesPerRadian
+            * constants.kShoulderArmGearRatio
+        )
+        elbowArmEncoderPulses = (
+            trueElbowPos
+            * constants.kTalonEncoderPulsesPerRadian
+            * constants.kElbowArmGearRatio
+        )
+        wristArmEncoderPulses = (
+            trueWristPos
+            * constants.kTalonEncoderPulsesPerRadian
+            * constants.kWristArmGearRatio
+        )
+
+        self.elbowArm.set(
+            Falcon.ControlMode.Position,
+            elbowArmEncoderPulses,
+            self._elbowFF(Rotation2d(trueElbowPos)),
+        )
+        self.shoulderArm.set(
+            Falcon.ControlMode.Position,
+            shoulderArmEncoderPulses,
+            self._shoulderFF(Rotation2d(trueShoulderPos)),
+        )
+        self.wristArm.set(
+            Falcon.ControlMode.Position,
+            wristArmEncoderPulses,
+            self._wristFF(Rotation2d(trueWristPos)),
+        )
```

Comment:
Maybe I can't read but I swear I only saw the Cartesian interpolation and one bit of Jointspace throughout it, is there one that's just better or something?

Reply:
its only those special cases in which the interpolation should exist at those stages -- relatively simple given how we swap between the two